File: go/runner/render.go
package runner

import (
	"fmt"
	"strings"

	"github.com/varar-dev/varar/go/core"
)

// RenderFailure is a pure human-readable rendering of a step failure, anchored
// to the .md; it reuses the core diff payloads.
func RenderFailure(failure core.StepFailure, _source, path string) string {
	switch failure.Error.Kind {
	case core.SECellMismatch:
		lines := []string{fmt.Sprintf("Cell mismatch in %s:", path)}
		var failing []core.CellDiff
		for _, c := range failure.Error.Cells {
			if !c.Ok {
				failing = append(failing, c)
			}
		}
		if len(failing) == 0 {
			lines = append(lines, "  (no failing cells)")
		}
		for _, cell := range failing {
			lines = append(lines, fmt.Sprintf(
				"  line %d | column '%s' — expected: %q, actual: %q",
				cell.Span.StartLine, cell.Column, cell.Expected, cell.Actual,
			))
		}
		return strings.Join(lines, "\n")
	default:
		return failure.Error.Message()
	}
}

File: rust/runner/src/render.rs
//! Pure human-readable rendering of a step failure, anchored to the `.md`.
//! Port of `varar_runner.render.render_failure`; reuses the core diff payloads.

use varar_core::error::StepFailure;

pub fn render_failure(failure: &StepFailure, _source: &str, path: &str) -> String {
    let error = &failure.error;
    if let Some(cells) = error.as_cell_mismatch() {
        let mut lines = vec![format!("Cell mismatch in {path}:")];
        let failing: Vec<_> = cells.iter().filter(|c| !c.ok).collect();
        if failing.is_empty() {
            lines.push("  (no failing cells)".to_string());
        }
        for cell in failing {
            lines.push(format!(
                "  line {} | column '{}' — expected: {:?}, actual: {:?}",
                cell.span.start_line, cell.column, cell.expected, cell.actual
            ));
        }
        return lines.join("\n");
    }
    error.message()
}
